module: fix loading from global folders on Windows

richardlau opened the pull request:
##### Checklist
- [x] `make -j8 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows commit guidelines
##### Affected core subsystem(s)

module
##### Description of change

According to [the `module` documentation](https://nodejs.org/dist/latest-v6.x/docs/api/modules.html#modules_loading_from_the_global_folders):

```
Additionally, Node.js will search in the following locations:

    1: $HOME/.node_modules
    2: $HOME/.node_libraries
    3: $PREFIX/lib/node

Where $HOME is the user's home directory, and $PREFIX is Node.js's configured node_prefix.
```

Loading from `$PREFIX/lib/node` is broken on Windows as the code is calculating  `$PREFIX/lib/node` 
relative to `process.execPath`, but on Windows `process.execPath` is `$PREFIX\node.exe` whereas everywhere else `process.execPath` is `$PREFIX/bin/node` (where `$PREFIX` is the root of the
installed Node.js).

e.g. If Node.js is in `c:\node`, the code is adding `c:\lib\node` to the lookup path instead of `c:\node\lib\node`:
```
C:\node>set NODE_DEBUG=module

C:\node>node nosuchfile
MODULE 12780: looking for "C:\\node\\nosuchfile" in ["C:\\Users\\IBM_ADMIN\\.nod
e_modules","C:\\Users\\IBM_ADMIN\\.node_libraries","C:\\lib\\node"]
module.js:472
    throw err;
    ^

Error: Cannot find module 'C:\node\nosuchfile'
    at Function.Module._resolveFilename (module.js:470:15)
    at Function.Module._load (module.js:418:25)
    at Module.runMain (module.js:605:10)
    at run (bootstrap_node.js:420:7)
    at startup (bootstrap_node.js:139:9)
    at bootstrap_node.js:535:3

C:\node>
```

gibfahn commented:
cc/ @nodejs/platform-windows 

I feel like this would be a semver-patch change, as it's fixing the actual behaviour to match the docs.

CI: https://ci.nodejs.org/job/node-test-pull-request/4677/

jasnell commented:
Given past history, we need to be very careful with changes to the module loader logic. @nodejs/ctc ... thoughts on this?
/cc @nodejs/platform-windows

joaocgreis commented:
There was already some discussion of this in https://github.com/nodejs/node/issues/6434 , but no decision.

What about removing those 3 search locations altogether? As the doc says, "these are mostly for historic reasons", so perhaps it's time to see them gone. `NODE_PATH` can be used to achieve the same functionality explicitly. Was there any discussion around this in the past?

If not, I would rather see these paths changed to places that actually make sense on Windows (like `%APPDATA%`). If that change is still too big, this PR at least makes node behave as described in the docs, so I'm +1 here.

richardlau commented:
cc @nodejs/testing since this PR also adds a new testcase to cover the documented behaviour.

sam-github commented:
@joaocgreis Removing them is an even larger change to a system we are trying to not break! I doubt anyone has an appetite for adding new default paths. I don't think adding $PREFIX/lib/node into the default module resolution path would fly as a new feature, but @richardlau isn't proposing that. Its an existing documented feature, we just want it to work. We have a use-case for it, which is why we noticed its broken on Windows.

sam-github commented on lib/module.js:
```diff
@@ -619,7 +619,14 @@ Module._initPaths = function() {
+    prefixDir = path.resolve(process.execPath, '..', '..');
+  }
+
+  var paths = [path.resolve(prefixDir, 'lib', 'node')];
```
If I understand this correctly, on non-Windows we have

```
$PREFIX/
   bin/node
   lib/
```

and on Windows we have

```
$PREFIX\
  node.exe
  lib/
```

So the previous code was adding a path like `c:\Program Files\lib` to the default node search path.. which is just flat-out wrong?

I think the code above could benefit from a comment describing what its doing. Particularly if my guess above is wrong!

richardlau replied:
Yes, your understanding is correct. I have added some comments to the code to clarify.

richardlau commented:
Rebased and added a comment to the changed code as suggested by @sam-github.

I've also edited the PR description to (hopefully) make it clearer what problem this is trying to address with an example showing the current (incorrect) behavior.

joaocgreis commented:
This is a bugfix, but for the remote case someone was relying on the old behavior, should we make this semver-major?

gibfahn commented:
> This is a bugfix, but for the remote case someone was relying on the old behavior, should we make this semver-major?

It doesn't really seem possible to evaluate whether this will break people, so I guess yes.

sam-github commented:
I find it really, really hard to imagine anyone is relying on such a bizarre search path. If you had node in `c:/Program Files/node/`, would you accidentally depend on modules in a directory called lib somewhere near by? How would you even discover that such a directory was being used to load modules?

Calling this semver-major vs patch isn't a just a terminology quibble, major prevents it from being backported to 4.x and 6.x, which means those LTS releases can never get this feature working as documented and intended. The bug will exist for them forever.... and for everyone else until 8.x comes next year.

Maybe this should have started with a bug report, so it was more clear that there is a documented feature that isn't working, but at this point it seems like process overhead.

richardlau commented:
It's even less likely (but still not impossible) that people will be relying on it because you cannot use `npm` to install into a directory called `lib` (which will always install into a `node_modules` directory) -- The modules would have to be manually placed there. 

@sam-github In terms of bug report there is nodejs/node#6434 which correctly identified the problem but then confused the directory with npm's global prefix (which is not the same thing).

richardlau commented:
One suggestion is that if we are really concerned that someone might be relying on the current incorrect behavior would be to augment the search path with the correct location instead of replacing the existing incorrect one. e.g. in the example (`node.exe` in `C:\node`):
```
C:\node>node nosuchfile
MODULE 12780: looking for "C:\\node\\nosuchfile" in ["C:\\Users\\IBM_ADMIN\\.nod
e_modules","C:\\Users\\IBM_ADMIN\\.node_libraries","C:\\lib\\node"]
```
The code is searching:
1. C:\Users\IBM_ADMIN\.node_modules
2. C:\Users\IBM_ADMIN\.node_libraries
3. C:\lib\node

As it is right now, this PR would correct this to:
1. C:\Users\IBM_ADMIN\.node_modules
2. C:\Users\IBM_ADMIN\.node_libraries
3. C:\node\lib\node

I'd prefer this version of the fix to land in master. 
For backports we could do:

1. C:\Users\IBM_ADMIN\.node_modules
2. C:\Users\IBM_ADMIN\.node_libraries
3. C:\lib\node
4. C:\node\lib\node

So modules in the incorrect location would be picked ahead of those in the correct location.

richardlau commented:
Is there anything I can do to move this forward?

jasnell commented:
ping @nodejs/ctc

richardlau commented on test/parallel/test-module-loading-globalpaths.js:
```diff
@@ -0,0 +1,106 @@
+} else {
+  common.refreshTmpDir();
+
+  // copy node binary into a test $PREFIX
```
Done

richardlau commented on test/parallel/test-module-loading-globalpaths.js:
```diff
@@ -0,0 +1,106 @@
+  const runTest = (expectedString, env) => {
+    const child = child_process.execFileSync(testExecPath,
+      [ __filename, 'child' ],
+      { env: env }
```
Done

richardlau commented on test/parallel/test-module-loading-globalpaths.js:
```diff
@@ -0,0 +1,106 @@
+      [ __filename, 'child' ],
+      { env: env }
+    );
+    assert.equal(child.toString().trim(), expectedString);
```
Done (and earlier assertion too).

richardlau commented:
Rebased on to current master and addressed points raised by @bnoordhuis.

gibfahn commented:
This seems good to land unless anyone in the CTC has any objections today/tomorrow.

bnoordhuis commented on lib/module.js:
```diff
@@ -617,7 +617,15 @@ Module._initPaths = function() {
-  var paths = [path.resolve(process.execPath, '..', '..', 'lib', 'node')];
+  // $PREFIX/lib/node, where $PREFIX is the root of the Node.js installation
+  var prefixDir;
+  // process.execPath is $PREFIX\node.exe on Windows, otherwise $PREFIX/bin/node
```
Can you punctuate these too? :)

richardlau replied:
Done

bnoordhuis commented on test/parallel/test-module-loading-globalpaths.js:
```diff
@@ -0,0 +1,102 @@
+      child_process.execFileSync(testExecPath, [ __filename, 'child' ],
+                                 { encoding: 'utf8', env: env });
+    },
+    new RegExp('Cannot find module \'' + pkgName + '\'')
```
Microscopic nit: can you indent by four spaces here?  Continuations are indented twice to make them stand out.

richardlau replied:
Done. I've been running `make lint` before committing so it disappointing the indentation isn't being flagged.

gibfahn commented:
One last cc/ @nodejs/platform-windows in case anyone has any objections.

And a CI: https://ci.nodejs.org/job/node-test-commit/8112/ (mostly useful for the linter).

richardlau commented:
Not obvious to me why https://ci.nodejs.org/job/node-test-linter/7268/console failed. 

`make lint` was clean for me when run locally.

jasnell commented:
@richardlau ... is it possible that you missed adding the file on the commit?

gibfahn commented:
@richardlau the linter fails sometimes when it runs on some of the FreeBSD hosts, not sure why.

CI 3: https://ci.nodejs.org/job/node-test-commit/8117/

gibfahn commented on lib/module.js:
```diff
@@ -617,7 +617,16 @@ Module._initPaths = function() {
-  var paths = [path.resolve(process.execPath, '..', '..', 'lib', 'node')];
+  // $PREFIX/lib/node, where $PREFIX is the root of the Node.js installation.
+  var prefixDir;
+  // process.execPath is $PREFIX/bin/node except on Windows where it is 
```
Trailing whitespace on this line (breaks the linter locally for me).

richardlau replied:
Ah sorry. I know I ran the linter before (which is why this comment is no longer a single line as it was >80) but I must have forgotten afterwards. I won't be able to fix this until tomorrow (only have my tablet with me ATM).

richardlau commented:
@gibfahn Fixed trailing whitespace. Sorry about that.

```
-bash-4.2$ make lint
./node tools/eslint/bin/eslint.js --cache --rulesdir=tools/eslint-rules \
  benchmark lib test tools
Total errors found: 0
-bash-4.2$
```

gibfahn commented:
CI 4: https://ci.nodejs.org/job/node-test-commit/8140/

gibfahn commented:
Landed in https://github.com/nodejs/node/commit/92e7c93fce831650a0aea554074b0d0a8eb97f1c and https://github.com/nodejs/node/commit/055482c21e2df94446d9796949d1bb810e8331bf

MylesBorins commented:
Cherry-picked to both v4 and v6. I plan to do a maintenance release for v4 including this and a couple other bugs that have been caught

refack commented:
🎉🎉🎉🎉🎉🎉🎉🎉🎉
Ping @tniessen RE: global `node_modules`
